torch/csrc/jit/pickler.cpp

```diff
@@ -0,0 +1,490 @@
+#include <torch/csrc/jit/pickler.h>
+
+namespace torch {
+namespace jit {
+
+using ::c10::IValue;
+
+PicklerClass getClass(const std::string& string) {
+  if (string == "TensorID") {
+    return PicklerClass::TENSOR;
+  } else if (string == "IntList") {
+    return PicklerClass::INTLIST;
+  }
+  AT_ERROR("Unknown class name for unpickler");
+}
+
+const std::string& getClassName(PicklerClass cls) {
+  static const std::string tensor_class("TensorID\n");
+  static const std::string intlist_class("IntList\n");
+  switch (cls) {
+    case PicklerClass::TENSOR:
+      return tensor_class;
+    case PicklerClass::INTLIST:
+      return intlist_class;
+    default:
+      AT_ERROR("Unknown class for pickler");
+  }
+}
+
+const std::string& getModuleName() {
+  static const std::string module_name("__main__\n");
+  return module_name;
+}
+
+const std::vector<char>& Pickler::stack() {
+  return stack_;
+}
+
+void Pickler::start() {
+  pushOpCode(OpCode::PROTO);
+  pushUint8(2);
+
+  // All attributes get pushed into a list and their indices saved in the
+  // module def
+  pushOpCode(OpCode::EMPTY_LIST);
+  pushOpCode(OpCode::MARK);
+}
+
+void Pickler::finish() {
+  pushOpCode(OpCode::APPENDS);
+  pushOpCode(OpCode::STOP);
+}
+
+void Pickler::addIValue(const IValue& ivalue) {
+  // Check if reference ivalue has been saved before
+  const void* ivalue_ptr = getPointer(ivalue);
+  if (ivalue_ptr) {
+    auto memo_entry = memo_.find(ivalue_ptr);
+    if (memo_entry != memo_.end()) {
+      // This value has already been pushed, just do a BINGET
+      pushBinGet(memo_entry->second);
+      return;
+    }
+  }
+
+  if (ivalue.isTensor()) {
+    pushTensor(ivalue);
+  } else if (ivalue.isTuple()) {
+    pushTuple(ivalue);
+  } else if (ivalue.isDouble()) {
+    pushDouble(ivalue);
+  } else if (ivalue.isInt()) {
+    // TODO: use BININT1/BININT2/LONG if possible/necessary
```

Comment:
we should fix it sooner than later, i64 ids are pretty common

Reply:
Saving it for a follow up since this PR is getting big